python/cudf/cudf/_libxx/join.pyx

```diff
@@ -0,0 +1,150 @@
+from cudf._libxx.lib cimport *
+from cudf._libxx.table cimport Table, columns_from_ptr
+from cudf._libxx.includes.join cimport *
+from libcpp.vector cimport vector
+from libcpp.pair cimport pair
+
+from collections import OrderedDict
+
+cpdef join(Table lhs,
+           Table rhs,
+           object left_on,
+           object right_on,
+           object how,
+           object method,
+           bool left_index=False,
+           bool right_index=False
+           ):
+    """
+    Call libcudf++ join for full outer, inner and left joins.
+    Returns a list of tuples [(column, valid, name), ...]
+    """
+
+    cdef Table c_lhs = lhs
+    cdef Table c_rhs = rhs
+
+    num_inds_left = len(left_on) + left_index
+    num_inds_right = len(right_on) + right_index
+    cdef vector[int] left_on_ind
+    cdef vector[int] right_on_ind
+    left_on_ind.reserve(num_inds_left)
+    right_on_ind.reserve(num_inds_right)
+
+    cdef vector[pair[int, int]] columns_in_common
+
+    # Views might or might not include index
+    cdef table_view lhs_view
+    cdef table_view rhs_view
+
+    # the result columns are all the left columns (including common ones)
+    # + all the right columns (excluding the common ones)
+    result_col_names = []
```

Comment:
Do we know the size of this up front? If so lets allocate up front.

Reply:
I think there might be a way to reorganize things such that this is doable, but does it actually matter as long as this is just a pure python list? It's never converted to a vector so no `push_back`'s are used and the `appends` that are done are all O(1) I believe.

Reply:
Python lists are an array of PyObject* under the hood, so each time you run `append` Python needs to allocate new memory for the new size of the list and copy things into the new memory. It hides all of this but it can slow things down tremendously versus default initializing the list with `[None] * size`.

Reply:
```
In [6]: %%time
   ...:
   ...: size = int(1e7)
   ...: my_list = []
   ...:
   ...: for i in range(size):
   ...:     my_list.append(i)
   ...:
CPU times: user 1.75 s, sys: 474 ms, total: 2.23 s
Wall time: 2.23 s

In [7]: %%time
   ...:
   ...: size = int(1e7)
   ...: my_list = [None] * size
   ...:
   ...: for i in range(size):
   ...:     my_list[i] = i
   ...:
CPU times: user 1.19 s, sys: 115 ms, total: 1.31 s
Wall time: 1.31 s
```

Reply:
Done